src/main/scala/tsp/Model.scala

```diff
@@ -0,0 +1,68 @@
+package tsp
+
+import oscar.cbls.{CBLSIntVar, Objective, Store, setSum}
+import oscar.cbls.business.routing.invariants.global.RouteLength
+import oscar.cbls.business.routing.model.VRP
+import oscar.cbls.core.objective.CascadingObjective
+import oscar.cbls.lib.invariant.numeric.Sum2
+import oscar.cbls.lib.invariant.set.SetSum
+
+object Model {
+
+  def apply(problem: Problem): Model = {
+    new Model(problem)
+  }
+
+}
+
+class Model(val problem: Problem) {
+
+  // city 0 is considered as the depot
+  private val v                          = 1;
+  private val n: Int                     = 0 + problem.nCities
+  lazy val tsp                        = new VRP(new Store(), n, v, debug = false)
+  private val oscarIdToTSPId: Array[Int] = Array.tabulate(n)(oscarId => Math.max(0, oscarId - v))
+
+  lazy val distanceMatrix: Array[Array[Long]] =
+    Array.tabulate(n)(from => {
+      val fromId = from
+      Array.tabulate(n)(to => {
+        val toId = to
+        problem.distances(fromId)(toId)
+      })
+    })
+
+  // Invariant keeping the length of the tour
+  val routeLengthInvariant: CBLSIntVar =
+    RouteLength(tsp.routes, n, v, (from, to) => distanceMatrix(from)(to))(0)
+  // invariant keeping the number of unrouted nodes
+  val nUnroutedInvariant: SetSum = setSum(tsp.unrouted)
```

Comment:
Here you should use the cardinality invariant instead, since setSum will sum the values of the nodes (so if nodes 1 and 5 are unrouted, this would return 6 instead of 2)

```suggestion
  val nUnroutedInvariant = cardinality(tsp.unrouted)
```